File: zig/src/tools/eval_schema.zig
//! eval-schema — offline SchemaCleanr parity runner.
//!
//! Reads JSONL schema-cleaning ops from stdin and writes one JSON response per
//! op. The Python eval driver canonicalizes output before byte comparison.

const std = @import("std");
const zeroclaw = @import("zeroclaw");
const schema_api = zeroclaw.api.schema;

const EvalError = error{InvalidScenario};

pub fn main() !void {
    var gpa = std.heap.GeneralPurposeAllocator(.{}){};
    defer _ = gpa.deinit();
    const allocator = gpa.allocator();

    const input = try std.io.getStdIn().readToEndAlloc(allocator, 16 * 1024 * 1024);
    defer allocator.free(input);

    const stdout = std.io.getStdOut().writer();
    var lines = std.mem.splitScalar(u8, input, '\n');
    while (lines.next()) |raw_line| {
        const line = std.mem.trim(u8, raw_line, " \t\r\n");
        if (line.len == 0) continue;
        try runOp(allocator, line, stdout);
    }
}

fn runOp(allocator: std.mem.Allocator, line: []const u8, writer: anytype) !void {
    var parsed = try std.json.parseFromSlice(std.json.Value, allocator, line, .{});
    defer parsed.deinit();

    const op = getString(parsed.value, "op") orelse return EvalError.InvalidScenario;
    const schema_value = getField(parsed.value, "schema") orelse return EvalError.InvalidScenario;

    if (std.mem.eql(u8, op, "clean_for_gemini")) {
        var cleaned = try schema_api.cleanForGemini(allocator, schema_value);
        defer schema_api.freeJsonValue(allocator, &cleaned);
        try writeJsonResult(writer, op, cleaned);
    } else if (std.mem.eql(u8, op, "clean_for_anthropic")) {
        var cleaned = try schema_api.cleanForAnthropic(allocator, schema_value);
        defer schema_api.freeJsonValue(allocator, &cleaned);
        try writeJsonResult(writer, op, cleaned);
    } else if (std.mem.eql(u8, op, "clean_for_openai")) {
        var cleaned = try schema_api.cleanForOpenai(allocator, schema_value);
        defer schema_api.freeJsonValue(allocator, &cleaned);
        try writeJsonResult(writer, op, cleaned);
    } else if (std.mem.eql(u8, op, "clean_conservative")) {
        var cleaned = try schema_api.clean(allocator, schema_value, .Conservative);
        defer schema_api.freeJsonValue(allocator, &cleaned);
        try writeJsonResult(writer, op, cleaned);
    } else if (std.mem.eql(u8, op, "validate")) {
        try writer.writeAll("{\"op\":\"validate\",\"result\":{");
        schema_api.validate(schema_value) catch {
            try writer.writeAll("\"error\":\"InvalidSchema\"}}\n");
            return;
        };
        try writer.writeAll("\"ok\":true}}\n");
    } else {
        return EvalError.InvalidScenario;
    }
}

fn writeJsonResult(writer: anytype, op: []const u8, result: std.json.Value) !void {
    try writer.writeAll("{\"op\":");
    try std.json.stringify(op, .{}, writer);
    try writer.writeAll(",\"result\":");
    try writeJsonValue(writer, result);
    try writer.writeAll("}\n");
}

fn writeJsonValue(writer: anytype, value: std.json.Value) !void {
    switch (value) {
        .null => try writer.writeAll("null"),
        .bool => |inner| try writer.writeAll(if (inner) "true" else "false"),
        .integer => |inner| try writer.print("{d}", .{inner}),
        .float => |inner| try std.json.stringify(inner, .{}, writer),
        .number_string => |inner| try writer.writeAll(inner),
        .string => |inner| try std.json.stringify(inner, .{}, writer),
        .array => |array| {
            try writer.writeByte('[');
            for (array.items, 0..) |item, i| {
                if (i != 0) try writer.writeByte(',');
                try writeJsonValue(writer, item);
            }
            try writer.writeByte(']');
        },
        .object => |object| {
            try writer.writeByte('{');
            var first = true;
            var iterator = object.iterator();
            while (iterator.next()) |entry| {
                if (!first) try writer.writeByte(',');
                first = false;
                try std.json.stringify(entry.key_ptr.*, .{}, writer);
                try writer.writeByte(':');
                try writeJsonValue(writer, entry.value_ptr.*);
            }
            try writer.writeByte('}');
        },
    }
}

fn getField(value: std.json.Value, key: []const u8) ?std.json.Value {
    if (value != .object) return null;
    return value.object.get(key);
}

fn getString(value: std.json.Value, key: []const u8) ?[]const u8 {
    const field = getField(value, key) orelse return null;
    if (field != .string) return null;
    return field.string;
}

File: eval-tools/src/bin/eval-schema.rs
//! eval-schema — offline SchemaCleanr parity runner.
//!
//! Reads JSONL schema-cleaning ops from stdin and writes canonical JSONL.

use std::io::{self, Read, Write};

use anyhow::{Context, Result};
use serde_json::Value;
use zeroclaw_api::schema::{CleaningStrategy, SchemaCleanr};

fn main() -> Result<()> {
    let mut input = String::new();
    io::stdin().read_to_string(&mut input)?;
    let mut output = Vec::new();

    for raw_line in input.lines() {
        let line = raw_line.trim();
        if line.is_empty() {
            continue;
        }

        let op_value: Value = serde_json::from_str(line).context("invalid scenario JSON")?;
        let op = op_value
            .get("op")
            .and_then(Value::as_str)
            .context("op missing")?;
        let schema = op_value.get("schema").cloned().context("schema missing")?;

        match op {
            "clean_for_gemini" => {
                write_result(&mut output, op, SchemaCleanr::clean_for_gemini(schema));
            }
            "clean_for_anthropic" => {
                write_result(&mut output, op, SchemaCleanr::clean_for_anthropic(schema));
            }
            "clean_for_openai" => {
                write_result(&mut output, op, SchemaCleanr::clean_for_openai(schema));
            }
            "clean_conservative" => {
                write_result(
                    &mut output,
                    op,
                    SchemaCleanr::clean(schema, CleaningStrategy::Conservative),
                );
            }
            "validate" => {
                let result = match SchemaCleanr::validate(&schema) {
                    Ok(()) => serde_json::json!({ "ok": true }),
                    Err(_) => serde_json::json!({ "error": "InvalidSchema" }),
                };
                write_result(&mut output, op, result);
            }
            _ => anyhow::bail!("unknown schema op: {op}"),
        }
    }

    io::stdout().lock().write_all(&output)?;
    Ok(())
}

fn write_result(output: &mut Vec<u8>, op: &str, result: Value) {
    let value = serde_json::json!({"op": op, "result": result});
    output.extend_from_slice(canonical_dump(&value).as_bytes());
    output.push(b'\n');
}

fn canonical_dump(value: &Value) -> String {
    match value {
        Value::Null => "null".to_string(),
        Value::Bool(inner) => inner.to_string(),
        Value::Number(inner) => inner.to_string(),
        Value::String(inner) => serde_json::to_string(inner).unwrap(),
        Value::Array(items) => {
            let rendered: Vec<String> = items.iter().map(canonical_dump).collect();
            format!("[{}]", rendered.join(","))
        }
        Value::Object(object) => {
            let mut keys: Vec<&String> = object.keys().collect();
            keys.sort();
            let rendered: Vec<String> = keys
                .iter()
                .map(|key| {
                    format!(
                        "{}:{}",
                        serde_json::to_string(key).unwrap(),
                        canonical_dump(&object[*key])
                    )
                })
                .collect();
            format!("{{{}}}", rendered.join(","))
        }
    }
}
